### PR TITLE
Querier worker refactor step2

### PR DESCRIPTION
**Pull request by pstibrany**
**What this PR does**: This is the second (and last step) of querier worker refactoring (step 1: https://github.com/cortexproject/cortex/pull/3455).

In this PR, querier worker has been modifed as follows:
- `processor` interface is introduced. Processor is responsible for starting streaming operation, fetching requests, executing them (via provided handler), and handling response.
- There are two `processor` implementations: `frontendProcessor`, working with query-frontend, and `schedulerProcessor` working with scheduler. Querier worker decides which one to use based on provided config.
- `processorManager` manages goroutines associated with single grpc connection to frontend or scheduler. It can be told to start more or fewer goroutines, based on configuration and discovered frontends / schedulers.
- `querierWorker` receives DNS updates, and starts / stops processor managers for each new/removed address.

Configuration and documentation is changed only minimally. Previously introduced `-querier.scheduler-dns-lookup-period` option has been removed in favor of using single `-querier.dns-lookup-period` for both frontend and scheduler.

**Checklist**
- [x] Tests updated
- [x] Documentation added
- [ ] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

File: pkg/querier/worker/frontend_processor_test.go [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+		defer running.Store(false)
+
+		mgr.processQueriesOnSingleStream(ctx, cc, "test:12345")
+	}()
+
+	time.Sleep(50 * time.Millisecond)
+	assert.Equal(t, true, running.Load())
+
+	cancel()
+	test.Poll(t, 100*time.Millisecond, false, func() interface{} {
```
Comment (pracucci) [inline pkg/querier/worker/frontend_processor_test.go:39]: The 100ms polling timeouts in these tests may trigger false positives on slow machines. I would suggest increasing it to 1s.

_Same comment applies to `TestServeCancelStopsProcess`._

File: pkg/querier/worker/frontend_processor_test.go [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"go.uber.org/atomic"
+	"google.golang.org/grpc"
+
+	"github.com/cortexproject/cortex/pkg/util"
+	"github.com/cortexproject/cortex/pkg/util/test"
+)
+
+func TestRecvFailDoesntCancelProcess(t *testing.T) {
```
Comment (pracucci) [inline pkg/querier/worker/frontend_processor_test.go:17]: I'm missing a thing. How is it asserted that doesn't cancel process?

**Review by pracucci — APPROVED** [review]
Good job! Definitely a needed refactoring. The diff is quite hard to follow, so I checked out the actual implementation and compared with the one in `master`. The core logic has been left intact and I haven't noticed any critical issue. I left few comments/questions, I would be glad you to take a look before merging. Everything LGTM 👏

File: pkg/querier/worker/worker.go [hunk 3]
```diff
@@ -3,218 +3,246 @@ package worker
...
-		if err != nil {
-			// watcher.Next returns error when Close is called, but we call Close when our context is done.
-			// we don't want to report error in that case.
-			if servCtx.Err() != nil {
-				return nil
-			}
-			return errors.Wrapf(err, "error from DNS watcher")
-		}
+func (f *querierWorker) stopping(_ error) error {
+	err := services.StopManagerAndAwaitStopped(context.Background(), f.subservices)
```
[the remark this answers is not in the document]
Comment (pstibrany) [reply pkg/querier/worker/worker.go:147]: You're right, that will work better. Otherwise we may stop the pool while it's still being used.

File: pkg/querier/worker/worker.go [hunk 4]
```diff
@@ -3,218 +3,246 @@ package worker
...
 
-		if mgr, ok := w.managers[addr]; ok {
-			mgr.concurrentRequests(concurrentRequests)
+		if f.cfg.MatchMaxConcurrency {
+			concurrency = f.cfg.MaxConcurrentRequests / len(f.managers)
+
+			// If max concurrency does not evenly divide into our frontends a subset will be chosen
+			// to receive an extra connection.  Frontend addresses were shuffled above so this will be a
+			// random selection of frontends.
+			if index < f.cfg.MaxConcurrentRequests%len(f.managers) {
```
[the remark this answers is not in the document]
Comment (pstibrany) [reply pkg/querier/worker/worker.go:211]: Is that a useful comment? I can put it back, but don't see much value in it.

File: pkg/querier/worker/frontend_processor_test.go (hunk 2 above)
Comment (pstibrany) [reply pkg/querier/worker/frontend_processor_test.go:17]: The point of this test is to verify that frontend processor's `processQueriesOnSingleStream` keeps running even though it cannot receive any queries from the frontend.

File: pkg/querier/worker/frontend_processor_test.go [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+	mgr := newFrontendProcessor(cfg, nil, util.Logger)
+	running := atomic.NewBool(false)
+	go func() {
+		running.Store(true)
+		defer running.Store(false)
+
+		mgr.processQueriesOnSingleStream(ctx, cc, "test:12345")
+	}()
+
+	time.Sleep(50 * time.Millisecond)
```
[the remark this answers is not in the document]
Comment (pstibrany) [reply pkg/querier/worker/frontend_processor_test.go:35]: We're not waiting for `true` here, but checking if it keeps running and doesn't stop just because it cannot contact frontend.

[the remark this answers is not in the document]
Comment (pstibrany) [reply pkg/querier/worker/frontend_processor_test.go:35]: I've modified the test to make this more obvious.

File: pkg/querier/worker/worker.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (pracucci) [reply pkg/querier/worker/worker.go]: You may be right. Probably not useful.

**Comment by pracucci** [comment]
Thanks @pstibrany for your follow up. Still LGTM ;)

**Comment by pstibrany** [comment]
Thanks for your review @pracucci!
